### DISCUSSION TITLE
H1-5.0-E-G2 Generation 2 not supported?

### DISCUSSION THREAD

**Discussion by jHamshaw** [nathanmarlor/foxess_modbus#553] [Support]
Hi

New installation this week - I appear to have a G2 inverter - many functions missing when compared to G1 - most notably any form of forced discharge and the advanced scheduler functions. Fox support say new firmware will be released "very soon".

In the meantime I'm trying a ModBus connection with an  Elfin-EW11 all connected and fired up. All cables meter-checked for continuity.

The Gen 2 user manual is identical to the Gen 1 in terms of RS485 connectivity.

However, when trying to configure the integration I get error message :

Error connecting to your adapter or inverter. Ensure the adapter is properly configured and is correctly wired to your inverter (see the setup link above), then try again. Details: **Inverter model not supported:** ''

The error message detail seems to suggest that my Gen 2 inverter won't work with this ModBus integration either? Is this really the case??

I'll be mightily pi33ed off if the latest Fox hardware is so firmware-crippled as not to be as useful as the Gen 1 - I had no idea that there even was a Gen 2 available when I had my install done....

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/a0cc705a-54d2-4b05-84dc-fdafb35322e8)

**Comment by canton7** [comment]
Interesting! All other inverter models expose their model number (e.g. "H1-5.0-E"), and we use this to determine what sort of inverter we're talking to. It looks like your inverter is just telling us its model is "", which means we don't know what we're talking to.

Can you get yourself set up with [modpoll](https://www.modbusdriver.com/modpoll.html), and take a look at holding registers 30000-30015:

```
modpoll.exe -m tcp -0 -1 -t 4 -a 247 -r 30000 -c 15 192.168.1.150
```

**Reply by jHamshaw** [reply]
Hi and thanks for the reply - here's what I get :

C:\Users\admin\Downloads\modpoll-3.11\win>modpoll -m tcp -0 -1 -t 4 -a 247 -r 30000 -c 15 192.168.1.150
modpoll 3.11 - FieldTalk(tm) Modbus(R) Master Simulator
Copyright (c) 2002-2024 proconX Pty Ltd
Visit https://www.modbusdriver.com for Modbus libraries and tools.

Protocol configuration: MODBUS/TCP, FC3
Slave configuration...: address = 247, start reference = 30000 (PDU), count = 15
Communication.........: 192.168.1.150, port 502, t/o 1.00 s, poll rate 1000 ms
Data type.............: 16-bit register, output (holding) register table

-- Polling slave...
[30000]: 18481
[30001]: 11573
[30002]: 11824
[30003]: 11589
[30004]: 11591
[30005]: 12832
[30006]: 8224
[30007]: 8224
[30008]: 0
[30009]: 0
[30010]: 0
[30011]: 0
[30012]: 0
[30013]: 0
[30014]: 0

**Reply by canton7** [reply]
Right perfect thanks. So they've moved to packing two ascii chars into each register, probably because it didn't fit otherwise. That's `H1-5.0-E-G2     ` (with 5 trailing spaces), which matches what we expect. I'll update the integration to detect that encoding when I get the chance.

**Reply by canton7** [reply]
Give v1.10.2b1 a try.

**Reply by jHamshaw** [reply]
Thanks - that was quick! I'm running my HA instance in a Windows 11 Hyper-V container - not sure how to update it manually.... I've only used HACS to date - I've got the 1.10.2b1.zip file...

**Reply by canton7** [reply]
You can use HACS. Browse to the integration in HACS, Hamburger menu in the top right, Redownload, then select the "Show beta versions" toggle.

**Reply by jHamshaw** [reply]
Thanks again - updated to 1.10.2b1 - still getting "Inverter model not supported" as before....

**Reply by canton7** [reply]
Can you look in the log (Settings - System - Logs) -- there should be an error message.

**Reply by jHamshaw** [reply]
Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_controller.py:494
Integration: foxess_modbus ([documentation](https://github.com/nathanmarlor/foxess_modbus), [issues](https://github.com/nathanmarlor/foxess_modbus/issues))
First occurred: 17:56:21 (2 occurrences)
Last logged: 17:58:08

Autodetect: failed to connect to (tcp://192.168.1.150:502)
Traceback (most recent call last):
  File "/config/custom_components/foxess_modbus/modbus_controller.py", line 494, in autodetect
    raise UnsupportedInverterError(full_model)
custom_components.foxess_modbus.common.exceptions.UnsupportedInverterError: Inverter model not supported: ''

**Reply by jHamshaw** [reply]
This error originated from a custom integration.

Logger: custom_components.foxess_modbus.modbus_controller
Source: custom_components/foxess_modbus/modbus_controller.py:493
Integration: foxess_modbus (documentation, issues)
First occurred: 17:56:21 (2 occurrences)
Last logged: 17:58:08

Did not recognise inverter model '' ([18481, 11573, 11824, 11589, 11591, 12832, 8224, 8224, 0, 0, 0, 0, 0, 0, 0])

**Reply by canton7** [reply]
Did you definitely restart HA after updating (not just a quick reload, but a proper full restart)?

**Reply by jHamshaw** [reply]
Yes - just restarted it again to be sure - same errors.....

**Reply by canton7** [reply]
Ah no, I've spotted my mistake. Hold tight!

**Reply by canton7** [reply]
Right, give v1.10.2b2 a try

**Reply by jHamshaw** [reply]
Excellent - connected! Now waiting for auto-configured energy dashboard to populate - many thanks! Off out shortly so I'll leave it to "bake" until tomorrow - thanks again for your help!

**Reply by canton7** [reply]
Glad to hear!

**Reply by jHamshaw** [reply]
So - an update... Whist my G2 inverter is connected, there's no data flow. Everything showing as "unavailable" - but I suspect this is the salient issue now :

sensor.fox_ess_ha1_5_0_e_g2_connection_status

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/b0c5734d-e85a-4060-9836-ceaaeb65a06d)

No rush - I'm unlikely to look at this again until tomorrow now ;-)

**Reply by canton7** [reply]
Argh, right, they've gone and changed how it communicates. My guess is that they've scrapped the holding registers, but let's see.

Using modpoll / the "Read registers" service (Developer Tools -> Services -> "FoxESS - Modbus: Read Registers"), can you try reading:

* Input and holding register 10016
* Input and holding register 30016
* Input and holding register 11000
* Input and holding register 31000
* Input and holding register 32000
* Input and holding register 41000

(So 12 different reads in total). That should give us a starting point...

**Reply by canton7** [reply]
@TonyM1958 it looks like you started figuring out the G2 register map. Did you get far with it?

**Reply by TonyM1958** [reply]
Not really, I don't have access to a G2 model to do any testing. I use input registers instead of holding registers for my HA integration. Must admit, I assumed/hoped the registers map would be the same...!

One way to check might be to install my integration and see what sensors are available. You can just uninstall and install yours immediately after, if that helps?

**Reply by canton7** [reply]
@TonyM1958 I can see some registers labelled "G2" in your spreadsheet. Where did you get those from?

It looks like the G2 isn't responding to input registers at all (or at least, it didn't like reading input register 10016, but it was happy reading holding register 30000), so I suspect that won't help unfortunately!

**Reply by TonyM1958** [reply]
Looked back at the data I used. The file is here: https://github.com/TonyM1958/HA-FoxESS-Modbus/wiki/Data-Sheets,-Manuals-and-Videos#references

![image](https://github.com/nathanmarlor/foxess_modbus/assets/63789168/58d114f3-0694-40ec-967b-a6edb15284b2)

The Fox AIO H1 modbus 2023 is the one. These registers are highlighted in the file and the file is dated 2023. They don't exist on H1 and are different on KH, so I surmised that the highlighted registers are new and that they might relate to H1-G2. The addresses are shown in pink in the spread sheet as they have not been verified as I haven't come across a product where you can read these (yet). Equally, they could be related to a new version of firmware.

Worth noting the SoH and battery capacity are also included here, which are now popping up in the Fox App v2.0 so must originate somewhere. I haven't tried to track these down just yet.

**Reply by TonyM1958** [reply]
BTW - the file lists input and not holding registers, so it does not tally with the hypothesis that H1-G2 is like KH. The overlap in registers between this file and KH has been identified as PV3 and PV4, which aren't available on H1 or H1-G2 in any case. Confusing isn't it...

**Comment by jHamshaw** [comment]
Here you go Input, then holding for each:
 
**- Input and holding register 10016**

error: >-
  Error reading registers. Type: RegisterType.INPUT; start: 10016; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(132, 4,
  IllegalAddress)

error: >-
  Error reading registers. Type: RegisterType.HOLDING; start: 10016; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(131, 3,
  IllegalAddress)
 
**- Input and holding register 30016**

values:
  "30016": 13872
values:
  "30016": 13872

**- Input and holding register 11000**

error: >-
  Error reading registers. Type: RegisterType.INPUT; start: 11000; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(132, 4,
  IllegalAddress)

error: >-
  Error reading registers. Type: RegisterType.HOLDING; start: 11000; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(131, 3,
  IllegalAddress)

**- Input and holding register 31000**

values:
  "31000": 1920
values:
  "31000": 1927

**- Input and holding register 32000**

values:
  "32000": 0
values:
  "32000": 0

**- Input and holding register 41000**

values:
  "41000": 0
values:
  "41000": 0

**Reply by TonyM1958** [reply]
With a bit of luck, they might have done the same thing as they did on KH from a quick look through - removed the input registers and presented most of the data through holding registers. Have a look at the modbus registers in my integration - most of the ones I read are the holding registers rather than the input registers with a few exceptions? (https://github.com/TonyM1958/HA-FoxESS-Modbus/blob/main/custom_components/HA-FoxESS-Modbus/modbusKH_RS485.yaml)

I get the inverter model number from 30000, serial number from 10008 and version info starting at 30016. I thought you did the same, but it sounds like you might have been using the input registers?

**Reply by TonyM1958** [reply]
Makes me wonder if H1-G2 is a stripped down KH inverter inside. It may not be a coincidence that KH and H1-G2 both don't have released support for Force Discharge yet?

**Reply by canton7** [reply]
Hmm that's a thought.

@jHamshaw, do you have a Back-up work mode available, in the app or on the front panel?

There are also some key differences in the register maps between H1 and KH - I'll dig them out when I'm back at a computer.

**Reply by jHamshaw** [reply]
No - just these three work modes, but a firmware update to enable missing stuff is due “very soon” according to Fox. Having seen the Octopus Agile integration coming in Fox app v2.0 I’m hoping this is this week in line with the app release….

![IMG_4854](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/b1b8b022-7b6c-4c4e-9a05-4093e7dcc36c)

**Reply by canton7** [reply]
Ooh OK, looks a lot like the new KH then.

Can you also try:

* 31016, 31053, 31054 (the KH moved from a single register to a register pair for load_power, and I wonder what the H1-G2 is doing)
* 44012 (the 1st gen H1 has a value here, but the KH doesn't)
* 31028, 31029 (where is the BMS connect state?)
* 32002 (the KH address for solar_energy_today, the H1 only has this as an input register)

**Reply by jHamshaw** [reply]
Input, then holding for each:

**31016**

values:
  "31016": 3083
values:
  "31016": 3067

**31053**
error: >-
  Error reading registers. Type: RegisterType.INPUT; start: 31053; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(132, 4,
  IllegalAddress)

error: >-
  Error reading registers. Type: RegisterType.HOLDING; start: 31053; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(131, 3,
  IllegalAddress)

**31054**
error: >-
  Error reading registers. Type: RegisterType.INPUT; start: 31054; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(132, 4,
  IllegalAddress)

error: >-
  Error reading registers. Type: RegisterType.HOLDING; start: 31054; count: 1;
  slave: 247 from tcp://192.168.1.150:502: Exception Response(131, 3,
  IllegalAddress)

**44012**
values:
  "44012": 0
values:
  "44012": 0

**31028**
values:
  "31028": 1
values:
  "31028": 1

**31029**
values:
  "31029": 1
values:
  "31029": 1

**32002**
values:
  "32002": 102
values:
  "32002": 102

**Reply by canton7** [reply]
OK thanks, so it's not entirely like the KH in that it hasn't moved to 32-bit values for some stuff, and it still seems to have the full set of remote control registers.

Also interesting that these are all exposed as input and holding registers: that's a departure from how things were done previously.

**Reply by canton7** [reply]
I'll start putting together my guess of what the register map might look like, then send that out as a beta. Rinse repeat a few times, and we should get close.

That's likely to be a process that'll take a little while I'm afraid: I tend to have limited time during the week.

**Reply by TonyM1958** [reply]
From some of the register values probed above, the old H1 LAN connection profile might be a good starting point to get the basic sensor information.

Ironic if Fox disabled the LAN port on H1 and then replicated the modbus mapping on H1-G2 though!

**Reply by canton7** [reply]
It does seem to have a bunch of things like the H1 LAN didn't have though...

My guess is that they've decided on a new architectural direction, and applied that to the new KH and H1-G2.

**Reply by TonyM1958** [reply]
yes, it definitely a super set

**Reply by TonyM1958** [reply]
You probably have better tools, but just in case this might help - this is how I found the registers for H3.

There is a zip file here that you can download and run in an HA terminal session: https://github.com/TonyM1958/HA-FoxESS-Modbus

This contains shell scripts that use modpoll to do some basic scanning for modbus registers. There are examples in this thread: https://github.com/TonyM1958/HA-FoxESS-Modbus/issues/7

The syntax is ./scan.sh \<type\> \<start\> \<step\> \<count\>

For example:

```
./scan.sh holding 0 1000 99
./scan.sh input 0 1000 99
```

Will scan for blocks of holding / input addresses and show where the inverter is responding. Once you have a block, you can zoom in and look at the values in that block:

```
./scan.sh holding 31000 1 100
```

Will step through 100 holding register starting at 31000 and show which ones respond.

This doesn't work for block data (where you have to read a number of registers) but it can quickly give you an idea of what might be out there.

Check the scripts for the connection it uses - you might need to adjust this, depending on your setup.

**Reply by jHamshaw** [reply]
I'll just leave this here - confirmation that H1 G2 firmware update is due at the end of the week.... ;-)

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/5f6793fc-eda5-4572-8cd6-ea31c9199c14)

**Reply by jHamshaw** [reply]
Still waiting for new firmware to be provided.......

**Reply by jHamshaw** [reply]
H1 Gen 2 firmware now updated to Ver 1.30 - I can see and set all of the new Work Mode and Octopus Agile stuff in Fox ESS Cloud. Nothing showing on the App Ver 2.0 for this yet though.

Tried removing and reinstalling Fox ESS - Modbus - just in case the new firmware made a difference - but I suspect not....

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/41aae6a5-f89b-4797-8e7b-48d7e6bc5b75)

**Reply by TonyM1958** [reply]
I've requested the G2 modbus map from Fox. This is currently under consideration and may short cut reverse engineering it.

As they released new firmware for both KH and G2 today, it supports my view that there is more in common  between these inverters than the names might suggest?

**Reply by canton7** [reply]
Ooh that'll be fantastic, thank you!

**Reply by TonyM1958** [reply]
@jHamshaw please can you run modpoll and reading holding registers 36001, 36002 and 36003 on your H1-G2 and let me know what you get back?

**Reply by jHamshaw** [reply]
Input and Holding return the same value for each :

values:
  "36001": 289

values:
  "36002": 258

values:
  "36003": 304

**Reply by TonyM1958** [reply]
Thanks - at least there is something there so that's progress. Those might be your inverter firmware versions. The closest sensible values I could get those to decode to might be Master 1.33, Slave 1.02 and Manager 1.48.

But that doesn't appear to agree with what you posted yesterday by the look of it?

**Reply by canton7** [reply]
The new KH seems to hex-encode the manager version now, which gives 1.30 as expected.

I believe the master and slave are just decimal, so 2.89 and 2.58? Does that match what the app says? Or maybe they've started hex-encoding those too, which gives 1.21 and 1.02.

**Reply by jHamshaw** [reply]
I had a firmware update applied late yesterday - versions showing now in Fox ESS cloud are :

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/9344a0dd-6ce5-4753-9c6b-82e929bd7dc9)

**Reply by canton7** [reply]
OK, they've switched to hex for all 3. Good to know!

**Reply by TonyM1958** [reply]
As you say, hex encoded by the look of it. Master 289 = 0x121 = 1.21, Slave 258 = 0x102 = 1.02, Manager 304 = 0x130 = 1.30. Wierd to use hex as the display rather than decimal but there you are...

Try reading a string (block of 16 words in hex) from 39002 and 39018 and then read single words from 39050 to 39052.

**Reply by jHamshaw** [reply]
values:
  "39002": 18481
  "39003": 11573
  "39004": 11824
  "39005": 11589
  "39006": 11591
  "39007": 12832
  "39008": 8224
  "39009": 8224
  "39010": 0
  "39011": 0
  "39012": 0
  "39013": 0
  "39014": 0
  "39015": 0
  "39016": 0
  "39017": 0

values:
  "39050": 0
  "39051": 2
  "39052": 2

**Reply by TonyM1958** [reply]
39002 - the first 8 words are 16 chars of an ASCII encoded byte string that reads H1-5.0-E-G2.
39108 - should be the serial number, probably encoded the same way?
39050 - model ID
39051 - number of string inputs
39052 - number of MPPT

That's about as far as I can go for now, unfortunately.

**Reply by jHamshaw** [reply]
39108 - should be the serial number, probably encoded the same way?

Or maybe not....

values:
  "39108": 0
  "39109": 0
  "39110": 0
  "39111": 0
  "39112": 0
  "39113": 0
  "39114": 0
  "39115": 0
  "39116": 0
  "39117": 0
  "39118": 0
  "39119": 37
  "39120": 0
  "39121": 0
  "39122": 0
  "39123": 0

**Reply by TonyM1958** [reply]
sorry, typo - 39018...

**Reply by jHamshaw** [reply]
39018 to 39025 each hold a 5-character number (39026 to 39033 are all 0) - I won't post my inverter serial number here for obvious reasons, but it's a 15-character alphanumeric string. I've reached my limits on how to decode the registers though....

**Reply by TonyM1958** [reply]
@jHamshaw PM me via the FB group (Tony Matthews) if you would like to see if we can get your inverter talking to HA?

**Comment by paultweedy77** [comment]
I'm going to have my H1-3.7-E swapped for a H1-5.0-E-G2 within the next few weeks. When that happens I'll be keen to get modbus working again very quickly, and will be on hand to prod the registers to try to create a new mapping, if Fox haven't generously supplied them themselves before that point..

**Reply by TonyM1958** [reply]
Thanks Paul - let me know when you have it up and running and we can talk through how to probe for registers if Fox doesn't come up with the goods.

**Reply by paultweedy77** [reply]
hi @TonyM1958 - My H1-5.0-E-G2 was installed today, up and running. Master 1.18, Slave 1.02, Manager 1.28. Currently digging around in the holding registers trying to make some sense of what's there.

**Reply by TonyM1958** [reply]
The registers we know about are in my integration already: https://github.com/TonyM1958/HA-FoxESS-Modbus

Follow the instructions in the wiki if you want to set this up.

**Reply by canton7** [reply]
Aha, I didn't realise you'd mapped the G2. I'll nab those for this integration soon.

**Reply by TonyM1958** [reply]
😂 I did drop you an email. Must have been a few weeks ago now. Guess you might not have seen it?

Its quite different to what you might expect with a few significant omissions compared with H1. Drop me an email if you have questions.

**Reply by canton7** [reply]
The last email I have is:

> I can't share the doc, but my best guess, still to be checked out by @jhamshaw, is here: https://github.com/TonyM1958/HA-FoxESS-Modbus/blob/main/custom_components/HA-FoxESS-Modbus/modbusH1G2_RS485_LAN.yaml

> Gives us a few days and hopefully we can verify this.

I waited a few days, then work got in the way again!

**Reply by paultweedy77** [reply]
I think I've completed the conversion of the entity mapping from Tony's integration, but my lack of Python foo is failing the final step - I can't get the entities to auto-create upon completion of the setup flow. I don't understand how the InverterModel class relates to the INVERTER_PROFILES mapping.

**Reply by canton7** [reply]
Ooh awesome, thank you!

`INVERTER_PROFILES` specifies a regex for each inverter, which is matched against the model number (and also pulls out the capacity, needed for the force charge/discharge stuff). 

During the setup flow, we read the model number, then find the inverter profile which matches, and store that (and the connection type and details) in the config.

On start-up, `sensor.py`/`number.py`/etc call `create_entities` which calls `inverter_connection_type_profile_from_config`, which pulls the `InverterModel` and `ConnectionType` out of the config, and uses that to find the `InverterModelConnectionTypeProfile` for that inverter + connection type. That gives us the `Inv` type and the `RegisterType`. These are then used when creating the entities (`entity_factory.create_entity_if_supported`), to select which entities are created and what address they have.

If you open a PR, I'm happy to take a look tomorrow.

**Reply by TonyM1958** [reply]
Always easier when you don't have the distraction of having to work for a living ;-)

Thanks to @jHamshaw for offering up his inverter as the guinea pig and putting up with a few iterations of trial and error. Hopefully, it will benefit a wider group of people in the long run.

**Reply by canton7** [reply]
I noticed some are marked as input registers (e.g. `foxess_inv1_eps_current_r`). Are those also readable as holding registers, or do we *need* to do a mix of input and holding reads?

**Reply by TonyM1958** [reply]
The main registers are all holding - apart from the inverter settings - so anything else is just copy and paste errors.

I've amended the files to remove the incorrect inputs. Some of them may read OK but there were outstanding questions I had about the accuracy of the EPS sensors that we didn't investigate as the main sensors were OK. Hopefully, testing against a wider base will throw up any other discrepancies.

**Reply by paultweedy77** [reply]
https://github.com/nathanmarlor/foxess_modbus/pull/585 for a first go.. 

Most entities that @TonyM1958 and @jHamshaw mapped seem to work.. a couple of nonsensical values (InvBatVolt 300V.. really? although I have just added a new ECS2900 slave so maybe that's expected)

Can't get charge periods or the charge period card to work. But at least I've got real-time data coming out..

**Comment by jHamshaw** [comment]
I’d recommend a firmware upgrade first - those versions are quite old.I’m currently on Master 1.21, Slave 1.02 and Manager 1.32 and£ there are still bugs in forced export and Agile integration - another update is due this week - I have a ticket raised and waiting….On 5 Apr 2024, at 16:12, Paul Tweedy ***@***.***> wrote:﻿
hi @TonyM1958 - My H1-5.0-E-G2 was installed today, up and running. Master 1.18, Slave 1.02, Manager 1.28. Currently digging around in the holding registers trying to make some sense of what's there.

**Comment by jHamshaw** [comment]
H1-5.0-E-G2 Inverter firmware >v1.32 (due imminently) is awaited to fix bugs - inability to set force discharge mode and in cloud Agile integration so could be linked…….On 6 Apr 2024, at 00:35, Paul Tweedy ***@***.***> wrote:﻿
#585 for a first go..
Most entities that @TonyM1958 and @jHamshaw mapped seem to work.. a couple of nonsensical values (InvBatVolt 300V.. really? although I have just added a new ECS2900 slave so maybe that's expected)
Can't get charge periods or the charge period card to work. But at least I've got real-time data coming out..

**Comment by canton7** [comment]
Thanks to @paultweedy77 and @TonyM1958, a first stab at support has just been released in v1.10.2b5. 

Please let me know if you find any issues with that pre-released.

Thanks!

**Comment by canton7** [comment]
@paultweedy77 Would you mind checking the _today and _total sensors (e.g. solar_energy_today)? The H3-Pro spec I have uses the same new registers as the G2, but the scale is 10x more than @TonyM1958's YAML file. So I'm wondering whether we accidentally have those values 10x too large.

**Reply by TonyM1958** [reply]
I'm reading the registers at 32000, 32002, same as H1. The new 32-bit energy registers with gain  x 100 are at 39601, 39603 etc, if that's the ones you mean? Those addresses didn't respond on H1_G2 when we were looking.

Lots of things in the modbus 'v2' spec didn't work on H1_G2. I think the spec is a goal rather than a statement of fact and H1-G2 is a hybrid / partial implementation. I can't rule out a firmware update that moves the modbus spec more towards v2 over time though??

**Comment by paultweedy77** [comment]
They both look ok I think, if I look at the history this week, after Friday (when my inverter was upgraded), they seem to be scaling in the same way as before:

<img width="1057" alt="Screenshot 2024-04-07 at 14 37 14" src="https://github.com/nathanmarlor/foxess_modbus/assets/46596533/1bc10b05-78f4-4412-9814-fcfefad694d6">

Slightly trickier to compare solar_generation_total as that reset with the new inverter but looks to be scaling in line with the generation:

<img width="1059" alt="Screenshot 2024-04-07 at 14 40 56" src="https://github.com/nathanmarlor/foxess_modbus/assets/46596533/3ad8b7d1-2778-469e-976a-623260e56f65">

Certainly doesn't look 10x out..

**Comment by paultweedy77** [comment]
An update on the large values I've been seeing on PV1 Power & PV2 Power registers on my new H1 G2..

Two random samples, seconds apart, just now:


```
[39282]: 	65528 (-8)
[39283]: 	0
.....
[39282]: 	65532 (-4)
[39283]: 	0

```
Also definitely present on the old 16 bit registers at the same time:
`
[31005]: 	65525 (-11)
`

They can occur at any time in the night, mainly on PV2 Power but sometimes on PV1 Power. I wonder if they're triggered by variations in marginal DC input after the sun has set waking up/shutting down an MPPT... More an annoyance than anything serious but definitely behaviour that wasn't present on my H1 G1.

<img width="699" alt="Screenshot 2024-04-07 at 20 21 45" src="https://github.com/nathanmarlor/foxess_modbus/assets/46596533/ee8ca88a-e435-4dea-b72a-3d41dabd29a5">

**Reply by canton7** [reply]
Aah, so looks like they defined it as a 2-register value, but only actually implemented one of the registers...
If we change the config to only read the lower register, thay should sort it out.

**Reply by paultweedy77** [reply]
Now the sun's out (I had the registers wrong in my comment above):

```
PV1:
[39279]: 	0
[39280]: 	586

PV2:
[39281]: 	0
[39282]: 	1797

```

**Reply by TonyM1958** [reply]
Looking like it's only a partial implementation of modbus v2 and they kept a lot of the modbus v1 registers. Have a look at:

[31000] PV1 Voltage
[31001] PV1 Current
[31002] PV1 Power
[31003] PV2 Voltage
[31004] PV2 Current
[31005] PV2 Power

These should read correctly as 16-bit int. Looks like they just copied the 16-bit power values to the new registers without extending them to 32-bits.

**Reply by paultweedy77** [reply]
Created https://github.com/nathanmarlor/foxess_modbus/pull/590 to remove upper registers for PV1 & PV2 power reads.

**Comment by paultweedy77** [comment]
Another difference discovered tonight - at least on firmware 1.18, the G2 doesn't support Back-up work mode.
You can set holding register 41000 to '2', but it does nothing. 
Confirmed that the inverter settings menu only contains Self Use, Feed-in First and Peak Shaving options.
@jHamshaw does yours look any different?

**Reply by canton7** [reply]
The new KH firmware is the same - although I think there was going to be a firmware update to fix that.

**Comment by jHamshaw** [comment]
I’m up to v1.32 already with another update due this week to fix issues with force discharge and Octopus Agile function in Fox ESS Cloud - neither of which work but at least appear now  - it all seems to be a bit of a “work in progress” still….Jim


On 7 Apr 2024, at 23:49, Paul Tweedy ***@***.***> wrote:﻿
Another difference discovered tonight - at least on firmware 1.18, the G2 doesn't support Back-up work mode.
You can set holding register 41000 to '2', but it does nothing.
Confirmed that the inverter settings menu only contains Self Use, Feed-in First and Peak Shaving options.
@jHamshaw does yours look any different?

**Reply by paultweedy77** [reply]
Agreed, definitely early days (again) with the G2. 
Back-up mode is easy enough to replicate by setting the discharge rate to 0A so not a biggie for me at the moment, just another ‘quirk’.

**Comment by jHamshaw** [comment]
On 7 Apr 2024, at 23:49, Paul Tweedy ***@***.***> wrote:﻿
Another difference discovered tonight - at least on firmware 1.18, the G2 doesn't support Back-up work mode.
You can set holding register 41000 to '2', but it does nothing.
Confirmed that the inverter settings menu only contains Self Use, Feed-in First and Peak Shaving options.
@jHamshaw does yours look any different?

**Reply by jHamshaw** [reply]
![IMG_4898](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/eaf31e05-632d-4339-aede-b2376bb3e209)

**Comment by paultweedy77** [comment]
Another potential G2 firmware difference. Not relevant to the integration but worth knowing anyway.

On my H1 G1, Max SoC used to limit the timed force charge from grid so it wouldn't go beyond what you set it to. Then it would dump any PV excess into the battery (under Self Use).

Now Max SoC seems to to the opposite; it doesn't limit the force charge from grid (my batteries charged to 100% last night despite Max SoC being set to 50% as I knew today would be sunny), but _does_ limit battery charging from PV; if current SoC is > Max SoC, any excess is exported, even under Self Use. If I set it above current SoC, the battery charges. 

Haven't tested Force Charge work mode yet (instead of time periods) to see if that behaves any differently.

**Reply by FozzieUK** [reply]
I’ve also had that reported about the KH as well - maxSoC doesn't limit grid charge, but does limit solar PV - I had been told this should be ‘fixed’ shortly in KH so this may only be a temporary ‘feature’ 😉

**Reply by TonyM1958** [reply]
it's not been working on H1 either since they introduced Force Charge work mode. The way I've heard most people are working around this is to set a schedule that spans the time of their cheap rate with Self Use and fdMin_Soc set to 50%, for example. If the battery is above the target SoC at the start of the period, it will be consumed and reduce grid usage. If the battery is lower than the target, it will slowly charge up and the house load will use the grid. Sounds simple, but it's a pretty effective way to get a target SoC and optimises the use of residual energy in the battery.

**Reply by canton7** [reply]
Max SoC on the H1 doesn't work if you use strategy periods, but it does still work with normal charge periods.

The problem there is not with Max SoC per se, but with the fact that strategy periods use remote control, and remote control doesn't respect Max SoC.

Also note that Force Charge as provided by this integration does respect Max SoC.

**Comment by jHamshaw** [comment]
On the basis of tickets I have open for my H1 G2  (unable to force discharge) I’m not sure Fox UK support even realise that as they don’t seem to understand how my current settings are even working (but they are!) I can't even set MaxSoC - for Force Charge or Self Use - I get an error code 40257-Parameter does not meet expectations…. 

I can't set a Force Discharge period at all - same error code...

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/130fa636-aa62-41f6-bcb9-f67af0d09fa6)

**Reply by jHamshaw** [reply]
Update - a full shutdown and restart of my inverter, BMS and PV in the (correct order) after V1.32 was installed has sorted my inability to set force discharge mode!

**Comment by jHamshaw** [comment]
Could that be an issue on my inverter (can’t set forced discharge) that needs fixing? Remote control = 1 in ModBus but can’t be changed on inverter front panel as nothing under that menu option…. No documentation that I can find either….

**Reply by canton7** [reply]
(Please respond using the gitlab web interface rather than by email: email replies don't follow the thread, and contain a ton of noise)

**Reply by canton7** [reply]
I'm not aware of any way to control the remote control using the front panel.

**Reply by jHamshaw** [reply]
Does anyone know what the Remote Control settings actually do?

![image](https://github.com/nathanmarlor/foxess_modbus/assets/68791717/46dfcd94-eb74-4378-a4c2-1af80d42395e)

**Reply by canton7** [reply]
Yes, and this integration uses those registers to provide intelligent "Force Charge" and "Force Discharge" work modes.

The fact that you're seeing those entities at all means that you're not using this integration.

**Reply by jHamshaw** [reply]
I'm currently using the TonyM1958 integration (I helped with registers etc. while he updated it at his request) as this one didn't work at all when I got my G2 inverter in February. Has it been updated to support H1 G2 (it's still not listed in supported models) now? Regardless of that, I still can't even set a force discharge period simply using Fox or ES apps either...."Write Error" ....

**Reply by TonyM1958** [reply]
The integration is read only for all registers so is just reporting the values. That's how I noticed Fox are using remote control for strategy periods, as the registers change to match your Force Discharge when it runs.

**Comment by canton7** [comment]
It looks like someone has Illegal Address trying to read the serial number on a G2: https://github.com/nathanmarlor/foxess_modbus/discussions/594#discussioncomment-9121946

Any ideas, anyone?

**Reply by canton7** [reply]
Wait a sec, why is it even trying to read 11000. I'll take a proper look when I'm at a computer.
